src/cluster.c

```diff
@@ -3602,6 +3721,226 @@ int clusterSendModuleMessageToTarget(const char *target, uint64_t module_id, uin
     return C_OK;
 }
 
+clusterMsgSendBlock *cloneMsgSendBock(clusterMsgSendBlock *msgblock)
+{
+    clusterMsgSendBlock *clone;
+    clone = zmalloc(sizeof(*clone));
+    memcpy(clone, msgblock, sizeof(*clone));
+    return clone;
+}
+
+listNode *findNode(clusterLink *link, clusterMsgSendBlock *msgblock)
+{
+    listNode *node;
+    for (node = listFirst(link->send_msg_queue); node; node = listNextNode(node))
+        if (node->value == msgblock)
+            return node;
+    return NULL;
+}
+
+int isAttachable(clusterLink *link, listNode *node)
+{
+    listNode *first, *last;
+
+    last = listLast(link->send_msg_queue);
+    if (last != node)
+        return 0;
+
+    first = listFirst(link->send_msg_queue);
+    if (first == last && link->head_msg_send_offset)
+        return 0;
+
+    return 1;
+}
+
+/* Returns true if messages could be attached to last message on the queue */
+int clusterAttachToPrevious(clusterLink *link, robj *channel, robj **messages,
```

Comment:
I might be missing something but this optimization seems to bring in lot of complexity. I would rather add a new block for each new MPUBLISH command.

Reply:
The idea is to combine multiple `PUBLISH` messages into a single `MPUBLISH` message if possible.  This will make existing code bases able to take advantage of the new efficiency upon upgrade -- without rewriting the client code.

Reply:
@hpatro OK, I will remove this optimization.  I was quite proud of it -- oh well.